stylegan_code_finder/networks/trans_u_net/utils.py

```diff
@@ -27,16 +44,66 @@ def _dice_loss(self, score, target):
         loss = 1 - loss
         return loss
 
-    def forward(self, inputs, target, weight=None, softmax=False):
+
+class TverskyLoss(LossFunction):
+    def __init__(self, n_classes: int, alpha: float, beta: float):
+        super().__init__(n_classes)
+        self.alpha = alpha
+        self.beta = beta
+
+    def _metric_calculation(self, score: torch.Tensor, target: torch.Tensor) -> float:
+        target = target.float()
+        smooth = 1e-5
+        tp = torch.sum(score * target)
+        fp = torch.sum((1-target) * score)
+        fn = torch.sum(target * (1-score))
+        loss = (tp + smooth) / (tp + self.alpha*fp + self.beta*fn + smooth)
+        loss = 1 - loss
+        return loss
+
+
+class HandwritingLoss(LossFunction):
+    def forward(self, inputs: torch.Tensor, target: torch.Tensor, weight=None, softmax: bool = False) -> float:
         if softmax:
             inputs = torch.softmax(inputs, dim=1)
         target = self._one_hot_encoder(target)
-        if weight is None:
-            weight = [1] * self.n_classes
         assert inputs.size() == target.size(), 'predict {} & target {} shape do not match'.format(inputs.size(),
                                                                                                   target.size())
-        loss = 0.0
-        for i in range(self.n_classes):
-            dice = self._dice_loss(inputs[:, i], target[:, i])
-            loss += dice * weight[i]
-        return loss / self.n_classes
+        # This presumes that the handwritten class is always the last
```

Comment:
This is a dangerous assumption to just put in a comment and not actually assert somewhere. Makes it hard to debug for someone who is not super familiar with the code.
However, if this class is removed one could test the assertion where the loss function is called.